### PR TITLE
Better caching of prototypes

### PR DESCRIPTION
**Pull request by odersky**
- Make IgnoredProto a cached type
 - Convert IgnroedProtos to WildcardTypes in wildApprox
 - Use `eql` instead of `equals` when hash-consing prototypes

### CODE REVIEW

**Comment by odersky** [comment]
test performance please

**Comment by odersky** [comment]
test performance please

File: compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala [hunk 1]
```diff
@@ -185,22 +196,27 @@ object ProtoTypes {
...
+        (name eq that.name) && memberProto.equals(that.memberProto) && (compat eq that.compat) && (privateOK == that.privateOK)
+      case _ =>
+        false
+
+    override def eql(that: Type): Boolean = that match {
+      case that: SelectionProto =>
+        (name eq that.name) && (memberProto eq that.memberProto) && (compat eq that.compat) && (privateOK == that.privateOK)
+      case _ =>
+        false
+    }
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala:219]: Does the slight difference between `equals` and `eql` have a semantic difference or only a performance difference? In any case, for maintainability, it is worth some explanation here.

**Review by liufengyun — APPROVED** [review]
LGTM

File: compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala (hunk 1 above)
Comment (odersky) [reply compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala:219]: It's only for performance.`eql` does a `eq` on all recursive calls, knowing that all elements are already hash-consed. I'll add an explanation in one of the next PRs

**Pull request merged**